File: vosk-prototype/Screw.py
from Tool import Tool
from ToolProperties import DiameterScrew
from ToolProperties import Length
import Data


class Screw(Tool):
    def __init__(self, tool_names, tool_handle):
        super().__init__(tool_names, tool_handle)

    def search_property(self, speech_transcription, tool_property_search):
        namefound = super(Screw, self).searchname(speech_transcription, tool_property_search)
        oldnamefoud = super(Screw, self).searcholdname(tool_property_search)
        if namefound or oldnamefoud:
            self.search_size(speech_transcription, tool_property_search)
            tool_property_search.is_tool_found_ = True

    def search_size(self, speech_transcription, tool_property_search):
        screw_spine_property_search = tool_property_search.screw_spine_property_search_
        if screw_spine_property_search.diameter_ != DiameterScrew.BadDiameter:
            return self.search_length(speech_transcription, tool_property_search)
        if screw_spine_property_search.length_ != Length.BadLength:
            return self.map_length_to_diameter(speech_transcription, tool_property_search)
        if screw_spine_property_search.diameter_ == DiameterScrew.BadDiameter and screw_spine_property_search.length_ == Length.BadLength:
            for screw_data in Data.ScrewSpineData:
                if speech_transcription.find(screw_data.value) != -1:
                    screw_spine_property_search.diameter_ = screw_data
                    screw_spine_property_search.screw_need_diameter_ = ""
                    self.search_length(speech_transcription, tool_property_search)
                    return
            for screw_data in Data.ScrewSpineData:
                for length in Data.ScrewSpineData[screw_data]:
                    if speech_transcription.find(length.value) != -1:
                        screw_spine_property_search.length_ = length
                        screw_spine_property_search.screw_need_length_ = ""
                        self.map_length_to_diameter(speech_transcription, tool_property_search)
                        return
            # we put index one :  because at the index one we put the name "screw" which is sufficient to perform the
            # next search for the length and diameter
            screw_spine_property_search.screw_need_diameter_ = self.tool_names[1]
            screw_spine_property_search.screw_need_length_ = self.tool_names[1]

    def search_length(self, speech_transcription, tool_property_search):
        screw_spine_property_search = tool_property_search.screw_spine_property_search_
        lengths = Data.ScrewSpineData.get(tool_property_search.screw_spine_property_search_.diameter_, False)
        if lengths:
            for length in lengths:
                if speech_transcription.find(length.value) != -1:
                    screw_spine_property_search.length_ = length
                    screw_spine_property_search.screw_need_length_ = ""
                    return
        # we put index one :  because at the index one we put the name "screw" which is sufficient to perform the
        # next search for the length
        screw_spine_property_search.screw_need_length_ = self.tool_names[1]

    def map_length_to_diameter(self, speech_transcription, tool_property_search):
        screw_spine_property_search = tool_property_search.screw_spine_property_search_
        for screwData in Data.ScrewSpineData:
            if speech_transcription.find(screwData.value) != -1:
                for length in Data.ScrewSpineData[screwData]:
                    if length == screw_spine_property_search.length_:
                        screw_spine_property_search.diameter_ = screwData
                        screw_spine_property_search.screw_need_diameter_ = ""
                        return
        # we put index one :  because at the index one we put the name "screw" which is sufficient to perform the
        # next search for the  diameter
        screw_spine_property_search.screw_need_diameter_ = self.tool_names[1]

File: Microsoft-prototype/Screw.cpp
#include "Screw.h"
#include "Data.h"
#include <iostream>
#include <iterator>
#include "StringData.h" 
using namespace std;


Screw::Screw(const std::vector<string>& toolNames, const ToolHandleType& toolHandle) :
	Tool(toolNames, toolHandle) {
}

void Screw::searchProperty(const string& speechTranscription, ToolPropertySearch& toolPropertySearch) const {
	const bool nameFound = Tool::searchName(speechTranscription, toolPropertySearch);
	const bool oldnameFound = Tool::searchOldName(toolPropertySearch);
	if (nameFound || oldnameFound) {
		searchSize(speechTranscription, toolPropertySearch);
		toolPropertySearch.isToolFound = true;
	}
}

void Screw::searchSize(const string& speechTranscription, ToolPropertySearch& toolPropertySearch)const {
	ScrewSpinePropertySearch& screwSpinePropertySearch = toolPropertySearch.screwSpinePropertySearch;
	if (screwSpinePropertySearch.diameter != DiameterScrew::BadDiameter) {
		return searchLength(speechTranscription, toolPropertySearch);
	}
	if (screwSpinePropertySearch.length != Length::BadLength) {
		return mapLengthToDiameter(speechTranscription, toolPropertySearch);
	}
	if (screwSpinePropertySearch.diameter == DiameterScrew::BadDiameter && screwSpinePropertySearch.length == Length::BadLength) {
		for (const auto& screwData : ScrewSpineData) {
			if (speechTranscription.find(diameterToString[screwData.first]) != string::npos) {
				screwSpinePropertySearch.diameter = screwData.first;
				screwSpinePropertySearch.screwNeedDiameter = "";
				searchLength(speechTranscription, toolPropertySearch);
				return;
			}
			for (const auto& length : screwData.second)
			{
				if (speechTranscription.find(lengthToString[length]) != string::npos) {
					screwSpinePropertySearch.length = length;
					screwSpinePropertySearch.screwNeedLength = "";
					mapLengthToDiameter(speechTranscription, toolPropertySearch);
					return;
				}
			}
		}
		screwSpinePropertySearch.screwNeedDiameter = toolNames_[1];
		screwSpinePropertySearch.screwNeedLength = toolNames_[1];
	}
}
void Screw::searchLength(const string& speechTranscription, ToolPropertySearch& toolPropertySearch) const {
	auto lengthsPtr = ScrewSpineData.find(toolPropertySearch.screwSpinePropertySearch.diameter);
	if (lengthsPtr != ScrewSpineData.end()) {
		for (const auto& length : lengthsPtr->second)
		{
			if (speechTranscription.find(lengthToString[length]) != string::npos)
			{
				toolPropertySearch.screwSpinePropertySearch.length = length;
				toolPropertySearch.screwSpinePropertySearch.screwNeedLength = "";
				return;
			}

		}
	}
	//we put index one : because at the index one, we put the name "screw" which is sufficient to perform the next search for the length
	toolPropertySearch.screwSpinePropertySearch.screwNeedLength = toolNames_[1];

}

void Screw::mapLengthToDiameter(const string& speechTranscription, ToolPropertySearch& toolPropertySearch)const {
	for (const auto& screwData : ScrewSpineData) {
		if (speechTranscription.find(diameterToString[screwData.first]) != string::npos) {
			if (std::find(screwData.second.begin(), screwData.second.end(), toolPropertySearch.screwSpinePropertySearch.length) != screwData.second.end()) {
				toolPropertySearch.screwSpinePropertySearch.diameter = screwData.first;
				toolPropertySearch.screwSpinePropertySearch.screwNeedDiameter = "";
				return;
			};
		}
	}
	//we put index one : because at the index one, we put the name "screw" which is sufficient to perform the next search for the Diameter
	toolPropertySearch.screwSpinePropertySearch.screwNeedDiameter = toolNames_[1];
}

